source/robot_model/src/Model.cpp

```diff
@@ -260,9 +262,138 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
   return this->forward_geometry(joint_state, frame_ids);
 }
 
-state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState&) const {
-  // TODO
-  return state_representation::JointPositions();
+Eigen::MatrixXd Model::cwln_weighted_matrix(const state_representation::JointPositions& joint_positions, const double& margin) {
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  for(int n=0; n < this->robot_model_.nq; n++){
+    double d=1;
+    if(joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n]+margin){
+      d = (this->robot_model_.lowerPositionLimit[n]+margin - joint_positions.get_positions()[n])/margin;
+    }
+    else if(this->robot_model_.upperPositionLimit[n] - margin < joint_positions.get_positions()[n]){
+      d = (joint_positions.get_positions()[n] - (this->robot_model_.upperPositionLimit[n]-margin))/margin;
+    }
+    if((joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n]) || (this->robot_model_.upperPositionLimit[n] < joint_positions.get_positions()[n])){
+      W_b(n,n) = 0;
+    }
+    else if((joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n]+margin) || (this->robot_model_.upperPositionLimit[n] - margin < joint_positions.get_positions()[n])){
+      W_b(n,n) = -2*d*d*d + 3*d*d;
+    }
+    else{
+      W_b(n,n) = 1;
+    }      
+  }
+  
+  return W_b;
+}
+
+Eigen::VectorXd Model::cwln_repulsive_potential_field(const state_representation::JointPositions& joint_positions, const double& margin) {
+  Eigen::VectorXd Psi(this->robot_model_.nq);
+  Eigen::VectorXd q = joint_positions.get_positions();
+  for(int i = 0; i < this->robot_model_.nq; i++) {
+    Psi[i] = 0;
+    if(q[i] < this->robot_model_.lowerPositionLimit[i] + margin) {
+      Psi[i] = this->robot_model_.upperPositionLimit[i] - margin - std::max(q[i], this->robot_model_.lowerPositionLimit[i]);
+    }
+    else if(this->robot_model_.upperPositionLimit[i] - margin < q[i]) {
+      Psi[i] = this->robot_model_.lowerPositionLimit[i] + margin - std::min(q[i], this->robot_model_.upperPositionLimit[i]);
+    }     
+  }
+  
+  return Psi;
+}
+
+void Model::clamp_joint_positions(state_representation::JointPositions& joint_positions) {
+  Eigen::VectorXd q_clamped = joint_positions.get_positions();
+  for(int i = 0; i < this->robot_model_.nq; i++) {
+    q_clamped[i] = std::min(std::max(q_clamped[i], this->robot_model_.lowerPositionLimit[i]), this->robot_model_.upperPositionLimit[i]);
+  }
+  joint_positions.set_positions(q_clamped);
+}
+
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                            const state_representation::JointState& current_joint_state,
```

Comment:
inconsistent continuation indent (just needs one more space)

Also, if you change the declaration in the header according to the comments, remember to change the signature in the definition here too